pep-0683.rst

```diff
@@ -0,0 +1,343 @@
+types, as well as objects exposed by the public C-API.
+
+Performance
+-----------
+
+A naive implementation shows a 4% slowdown. [benchmarks]_
+Several promising mitigation strategies will be pursued in the effort
+to bring it closer to performance-neutral.
+
+On the positive side, immortal objects save a significant amount of
+memory when used with a pre-fork model.  Also, immortal objects provide
+opportunities for specialization in the eval loop that would improve
+performance.
+
+Backward Compatibility
+-----------------------
+
+This proposal is completely compatible.  It is internal-only so no API
+is changing.
+
+The approach is also compatible with extensions compiled to the stable
+ABI.  Unfortunately, they will modify the refcount and invalidate all
+the performance benefits of immortal objects.  However, the high bit
+of the refcount will still match ``_Py_IMMORTAL_REFCNT`` so we can
+still identify such objects as immortal.
+
+No user-facing behavior changes, with the following exceptions:
+
+* code that inspects the refcount (e.g. ``sys.getrefcount()``
+  or directly via ``ob_refcnt``) will see a really, really large
+  value
+* ``Py_SET_REFCNT()`` will be a no-op for immortal objects
+
+Neither should cause a problem.
+
+Alternate Python Implementations
+--------------------------------
+
+This proposal is CPython-specific.
+
+Security Implications
+---------------------
+
+This feature has no known impact on security.
+
+Maintainability
+---------------
+
+This is not a complex feature so it should not cause much mental
+overhead for maintainers.  The basic implementation doesn't touch
+much code so it should have much impact on maintainability.  There
+may be some extra complexity due to performance penalty mitigation.
+However, that should be limited to where we immortalize all
+objects post-init and that code will be in one place.
+
+Non-Obvious Consequences
+------------------------
+
+* immortal containers effectively immortalize each contained item
+* the same is true for objects held internally by other objects
+  (e.g. ``PyTypeObject.tp_subclasses``)
+* an immortal object's type is effectively immortal
+* though extremely unlikely (and technically hard), any object could
+  be incref'ed enough to reach ``_Py_IMMORTAL_REFCNT`` and then
+  be treated as immortal
+
+??? immortal != immutable
+
+
+Concerns
+========
+
+(This topic has been discussed previously on python-dev. [python-dev_])
+
+Concerns have centered around the performance penalty.
+Mitigation is discussed above.
+
+
+Specification
+=============
+
+The approach involves these fundamental changes:
+
+* add ``_Py_IMMORTAL_REFCNT`` (the magic value) to the internal C-API
+* update ``Py_INCREF()`` and ``Py_DECREF()`` to no-op for objects with
+  the magic refcount (or its most significant bit)
+* do the same for any other API that modifies the refcount
+* ensure that all immortal objects are cleaned up during
```

Comment:
How would this work? Some list of pointers that you can iterate through to free at runtime finalization?

Reply:
I'll have to check with Eddie, but I expect the plan was to walk through the GC objects.

Reply:
@brettcannon at a high level, there are three different kinds of cases that we need to care about when immortalizing objects to make sure that we correctly clean them up:

1) Runtime globals: Py_None, Py_True, Py_False, Small Ints, and types defined by: `PyTypeObject Foo =`. None of these instances are not cleaned up today, thus, the behavior will remain the same even after they are made immortal.

2) Immortalizing non-containers during runtime: Examples of these include interned strings. All of these objects have to be found during runtime shutdown, so they must be tracked somewhere. Fortunately, the biggest use case of this would be interned strings which are already tracked in a python dictionary. We should just refer to those at runtime shutdown to properly clean them up. But yes, the generic strategy is to track and deallocate at shutdown.

3) Immortalizing containers during runtime: Example of these include immortalizing the heap after initialization. To correctly track these instances, we'll leverage the GC's permanent generation by pushing all immortalized containers there. During runtime shutdown, the strategy will be to first let the runtime try to do its best effort of deallocating these instances normally. Most of the module deallocation will now be handled by `pylifecycle.c:finalize_modules` which cleans up the remaining modules as best as we can. It will change which modules are available during `__del__` but that's already defined as undefined behavior by the [docs](https://docs.python.org/3/reference/datamodel.html?highlight=__del__#object.__del__). Optionally, we could do some topological disorder to guarantee that user modules will be deallocated first before the stdlib modules. Finally, anything leftover (if any) can be found through the permanent generation gc list which we can clear after `finalize_modules`.

Reply:
> 1. None of these instances are cleaned up today, thus, the behavior will remain the same even after they are made immortal.

Does it make sense to continue to do that, or should we fix the situation?

Reply:
Static objects should not be freed.